doc/adrs/041_add_support_for_tools_preconditions.adoc

```diff
@@ -0,0 +1,87 @@
+
+Remove `toolSections` retrieving part from `getDiagramDescriptionQuery` and add it on `ContextualPalette`` openning.
+
+=== GraphQL API
+
+`ContextualPalette` should ask for the `toolsSections` while openning with a `useQuery<..., ...> ` thanks to the following GraphQL query :
+
+```
+query getToolSections($editingContextId: ID!, $diagramId: ID!, $diagramElementId: ID!) {
+  viewer {
+    editingContext(editingContextId: $editingContextId) {
+      representation(representationId: $diagramId) {
+        description {
+          ... on DiagramDescription {
+            toolSections(diagramElementId: $diagramElementId) {
+            // All the tool sections and tool fields such as id, label, imageURL etc
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+=== Backend
+
+Only the relevant tools will be return by this data fetcher, the `DiagramDescriptionToolSectionsDataFetcher, heavily inspired by the Sirius Web `EditingContextQueryBasedObjectDataFetcher`` with its custom internal input and payload.
+
+It will need to use the `IEditingContextEventProcessorRegistry`` to ask for the tool sections since you will need to be in the thread of the editing context to have the semantic element to run the precondition.
+
+```
+package o.e.s.c.collaborative.diagrams.handlers
+
+public ToolSectionEventHandler implements IDiagramEventHandler {
+
+  private final IDiagramQueryService diagramQueryService;
+  private final IObjectService objectService;
+  private final List<IToolSectionProvider> toolSectionProviders;
+
+  public void handle(One<IPayload> payloadSink, Many<ChangeDescription> changeDescriptionSink, IEditingContext editingContext, IInput input) {
+      // Implentation temporaire
+     return this.representationDescriptionSearchService.findById(this.diagram.getDescriptionId()).map(diagramDescription -> diagramDescription.getToolSections());
+
+     ToolSectionInput toolSectionInput = (ToolSectionInput) input;
+     var optionalTargetElement = this.diagramQueryService.findNodeById(input.getDiagramElementId()).flatMap(node -> this.ObjectService.findById(node.getTargetId()));
+     // Find the tool section providers which canHandle the diagram
+     // Ask this provider for the tool
+  }
+}
+```
+
+
+=== Compatibility layer
+
+We will probably need a `IToolSectionProvider`` to compute the preconditions for the compatibility layer.
+
+```
+public interface IToolSectionProvider {
+  boolean canHandle(..);
+  List<ToolSection> handle(Object targetElement,...);
+}
+```
+
+The code of both `ViewConverter.createToolSections(AQLInterpreter)` and `ToolProvider.getToolSections(Map<UUID, NodeDescription>, List<EdgeDescription>, DiagramDescription, List<Layer>)` would thus be transferred to two `IToolSectionProvider`
+
+The precondition of the tool will be evaluated by the `IToolSectionProvider` dedicated to compatibility layer (with the old code from `ToolProvider#getToolSections`)
+
```

Comment:
If we start supporting tools' preconditions in the API, it should also be supported in Views:
* add a `preconditionExpression` to `org.eclipse.sirius.components.view.Tool`
* update `org.eclipse.sirius.components.emf.view.ViewConverter.createToolSections(AQLInterpreter)` to pass the precondition to the Node/Edge tools it creates.